File: Sphere Online Judge/Longest_Palindromic_Substring.py
# SPOJ - LPS - Longest Palindromic Substring
# https://www.spoj.com/problems/LPS/

N = int(input())
string = input()

def manachers_algorithm(s):
    if not s:
        return ""
    
    t = "$#" + "#".join(s) + "#^"
    
    p = [0]*len(t)
    c = d = 1
    
    for i in range(2, len(t)-1):
        mirror = c-(i-c)
        p[i] = max(0, min(p[mirror], d-i))
        while(t[i-1-p[i]] == t[i+1+p[i]]):
            p[i] += 1
        if i + p[i] > d:
            c = i
            d = i + p[i]

    (k, i) = max((p[i], i) for i in range(1, len(p)-1))
    start = (i - k) // 2
    end = (i + k) // 2
    
    return s[start:end]
    
print(len(manachers_algorithm(string)))

File: Sphere Online Judge/Longest_Palindromic_Substring.cpp
// SPOJ - LPS - Longest Palindromic Substring
// https://www.spoj.com/problems/LPS/

#include "bits/stdc++.h"
using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int N; cin >> N;
    string s; cin >> s;

    string t = "$#";
    for (char c: s) {
        t.push_back(c);
        t.push_back('#');
    }
    t.push_back('^');
    vector<int> p(t.size());
    int c = 1, d = 1;

    for(int i = 2; i < t.size()-1; ++i){
        int mirror = c-(i-c);
        p[i] = max(0, min(p[mirror], d-i));
        while (t[i-p[i]-1] == t[i+p[i]+1]) {
            p[i]++;
        }
        if (i+p[i] > d) {
            c = i;
            d = i+p[i];
        }
    }

    int center = 1, radius = p[1];
    for(int i =1; i < t.size()-1; ++i){
        if (p[i] > radius) {
            radius = p[i];
            center = i;
        }
    }

    int start = (center-radius) / 2;
    int end = (center+radius) / 2;

    cout << end-start << endl;
}
